File: frontend/src/types/product.ts
export interface Category {
  id: string;
  name: string;
  parent_id: string | null;
  depth?: number;   // уровень вложенности (0 — корень) для показа деревом
}

export interface Product {
  id: string;
  name: string;
  description: string | null;
  image_url: string | null;
  images: string[];
  attributes: { name: string; value: string }[] | null;
  article: string | null;
  price: string;
  stock: number;
  available: boolean;
  is_active: boolean;
  // Слаги активных промо-категорий товара (по возрастанию display_order; первый — «основной»).
  promo_slugs?: string[];
  // «Старая цена» для показа зачёркнутой (уже готова к показу; null — не показываем).
  old_price?: string | null;
  // Deprecated: выводятся из membership на переходный период — новый код использует promo_slugs.
  is_new?: boolean;
  is_sale?: boolean;
  is_hot?: boolean;
  category: Category | null;
  updated_at: string;
  // Товар с обязательной маркировкой «Честный знак» (вычисляется на границе API).
  chestnyZnak?: boolean;
}

// Промо-категория для витрины (меню, ленты главной, страницы разделов).
export interface PromoCategory {
  slug: string;
  title: string;
  subtitle: string | null;
  icon: string | null;         // имя файла в media_storage; URL — /api/v1/admin/media/<icon>
  display_order: number;
  show_in_menu: boolean;
  show_on_home: boolean;
  show_old_price: boolean;
}

export interface ProductListResponse {
  items: Product[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

File: backend/app/schemas/product.py
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, model_validator
from app.core.timeutil import UtcDatetime


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    parent_id: str | None = None
    depth: int = 0  # уровень вложенности (0 — корень), вычисляется из кода в названии
    icon: str | None = None  # иконка категории (media_storage); свойство самой категории, None — без иконки


class TopCategoryOut(BaseModel):
    """Один слот блока «Топ категорий» на главной (уже резолвленный для витрины).

    Хранит только порядок (см. :mod:`app.services.top_categories`); ``name`` и ``icon`` берутся
    из самой категории по ``category_id`` — иконка не дублируется в конфиге блока.
    """
    category_id: str
    name: str            # актуальное имя категории из БД (следует за синхронизацией МойСклад)
    icon: str | None = None  # иконка КАТЕГОРИИ (Category.icon); None — плитка без картинки


class ProductOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None = None
    image_url: str | None = None          # первая картинка (для карточки/списка/OG)
    images: list[str] | None = None       # все картинки (для галереи на карточке товара)
    attributes: list[dict] | None = None  # характеристики из МойСклад: [{name, value}]
    article: str | None = None
    price: Decimal
    stock: int
    available: bool = True
    is_active: bool
    # Слаги активных промо-категорий товара (по возрастанию display_order; первый — «основной»).
    promo_slugs: list[str] = []
    # «Старая цена» для показа зачёркнутой — уже с учётом наценки/скидки и только когда её нужно
    # показывать (см. _product_out); None — не показываем.
    old_price: Decimal | None = None
    # Deprecated: выводятся из membership на переходный период (пока фронт не перешёл на promo_slugs).
    is_new: bool = False        # = принадлежит категории со слагом «novinki»
    is_sale: bool = False       # = слаг «special»
    is_hot: bool = False        # = слаг «hot»
    category: CategoryOut | None = None
    updated_at: UtcDatetime

    @model_validator(mode="after")
    def _normalize_images(self):
        # Всегда отдаём список: если images пуст, но есть image_url — кладём его одним.
        self.images = self.images or ([self.image_url] if self.image_url else [])
        return self


class ProductListOut(BaseModel):
    items: list[ProductOut]
    total: int
    page: int
    page_size: int
    pages: int
